File: programs/vote-aggregator/packages/tests/src/TestData/clan/configureClan.ts
import {Keypair, PublicKey} from '@solana/web3.js';
import {ClanTestData} from '../../VoteAggregator';
import {buildKeypair} from '../..';
import BN from 'bn.js';

export type ConfigureClanTestData = {
  clan: ClanTestData & {root: PublicKey};
  clanAuthority: Keypair | 'owner' | 'delegate';
  newDelegate?: PublicKey;
  newName?: string;
  newDescription?: string;
  newMinVotingWeightToJoin?: BN;
  newAcceptTemporaryMembers?: boolean;
  error?: string;
};

export const configureClanTestData: ConfigureClanTestData[] = [
  {
    clan: {
      root: new PublicKey('7yM5fXw8M64eF73SXqnU4Cbs1QsNUM2sRiwCjvmropJ6'),
      address: new PublicKey('4vK9oDnDgPbFGtCwdJvm3hHay9EueHDxJfNoSZMnQEZw'),
      owner: buildKeypair(
        '4NGov9AUKpbbkQU9AaS5ZqhEEzgWG3MSZrziDsLuWQBa',
        [
          144, 65, 137, 187, 226, 210, 92, 164, 117, 223, 216, 213, 155, 153,
          19, 240, 204, 28, 237, 169, 169, 15, 103, 217, 110, 21, 59, 71, 241,
          13, 203, 10, 50, 6, 34, 9, 6, 208, 204, 76, 233, 204, 56, 22, 103,
          175, 32, 153, 194, 176, 105, 137, 18, 68, 226, 213, 219, 48, 139, 32,
          54, 59, 134, 201,
        ]
      ),
      name: 'Marinade',
      description: 'Marinade is a test clan',
      size: 1000,
    },
    clanAuthority: 'owner',
    newDelegate: new PublicKey('Fz9TQxgjdERVNt7hP8DcQQwvM4oXL1oaGSbL1jX6HvJz'),
    newName: 'LIDO',
    newDescription: 'No more marinade',
    newMinVotingWeightToJoin: new BN(100),
    newAcceptTemporaryMembers: false,
  },
];

File: programs/vote-aggregator/programs/vote-aggregator/src/processor/clan/configure_clan.rs
use anchor_lang::prelude::*;

use crate::error::Error;
use crate::events::clan::{
    ClanAcceptTemporaryMembersChanged, ClanDelegateChanged, ClanDescriptionChanged,
    ClanMinVotingWeightToJoinChanged, ClanNameChanged,
};
use crate::state::Clan;

#[derive(Accounts)]
pub struct ConfigureClan<'info> {
    #[account(mut)]
    clan: Account<'info, Clan>,

    #[account(
        constraint = clan_authority.key() == clan.owner ||
            clan_authority.key() == clan.delegate
        @ Error::WrongClanAuthority,
    )]
    clan_authority: Signer<'info>,
}

impl<'info> ConfigureClan<'info> {
    pub fn set_delegate(&mut self, new_delegate: Pubkey) -> Result<()> {
        let old_delegate = self.clan.delegate;
        self.clan.delegate = new_delegate;
        if new_delegate != old_delegate {
            emit!(ClanDelegateChanged {
                clan: self.clan.key(),
                old_delegate,
                new_delegate,
            });
        }
        Ok(())
    }

    pub fn set_name(&mut self, new_name: String) -> Result<()> {
        let old_name = self.clan.name.clone();
        self.clan.name = new_name.clone();
        if new_name != old_name {
            emit!(ClanNameChanged {
                clan: self.clan.key(),
                old_name,
                new_name,
            });
        }
        Ok(())
    }

    pub fn set_description(&mut self, new_description: String) -> Result<()> {
        let old_description = self.clan.description.clone();
        self.clan.description = new_description.clone();
        if new_description != old_description {
            emit!(ClanDescriptionChanged {
                clan: self.clan.key(),
                old_description,
                new_description,
            });
        }
        Ok(())
    }

    pub fn set_min_voting_weight_to_join(
        &mut self,
        new_min_voting_weight_to_join: u64,
    ) -> Result<()> {
        let old_min_voting_weight_to_join = self.clan.min_voting_weight_to_join;
        self.clan.min_voting_weight_to_join = new_min_voting_weight_to_join;
        if new_min_voting_weight_to_join != old_min_voting_weight_to_join {
            emit!(ClanMinVotingWeightToJoinChanged {
                clan: self.clan.key(),
                old_min_voting_weight_to_join,
                new_min_voting_weight_to_join,
            });
        }
        Ok(())
    }

    pub fn set_accept_temporary_members(
        &mut self,
        new_accept_temporary_members: bool,
    ) -> Result<()> {
        let old_accept_temporary_members = self.clan.accept_temporary_members;
        self.clan.accept_temporary_members = new_accept_temporary_members;
        if new_accept_temporary_members != old_accept_temporary_members {
            emit!(ClanAcceptTemporaryMembersChanged {
                clan: self.clan.key(),
                old_accept_temporary_members,
                new_accept_temporary_members,
            });
        }
        Ok(())
    }
}
